### PR TITLE
Add analog support for RP2040

### PR DESCRIPTION
**Pull request by sigprof**
## Description

Add analog support for RP2040.

The low-level driver is already present in ChibiOS-Contrib, but the fix from ChibiOS/ChibiOS-Contrib#355 is required to make it actually work; without that fix the code will compile, but `adc_read()` will work only once (on the second call it will lockup).

TODO:
- [x] Add `RP_IRQ_ADC1_PRIORITY` to the default `mcuconf.h` files.
- [x] Update documentation for the supported pins and driver configuration options.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by KarlK90** [comment]
With https://github.com/qmk/qmk_firmware/pull/17915 being merged to `develop` the ADC fixes are now available in QMK :tada:.

**Review by tzarc — APPROVED** [review]

**Comment by sigprof** [comment]
The joystick code also builds and works with the fix from #19602.

The temperature sensor could be read with `adc_read(TO_MUX(4, 0))`, but the results are not really precise (the datasheet says that the precision greatly depends on the reference voltage, which in most cases is generated by some LDO not intended to be used as an analog reference).  However, the sensor needs to be enabled by `adcRPEnableTS(&ADCD1);`, and that call works only after the ADC initialization (e.g., after a dummy `adc_read()`).  Because the sensor has some non-zero power consumption, and probably nobody would actually want to use it, that initialization is left to the user.

File: keyboards/handwired/onekey/rp2040/mcuconf.h [hunk 1]
```diff
@@ -10,3 +10,9 @@
 
 #undef RP_PWM_USE_PWM4
 #define RP_PWM_USE_PWM4 TRUE
+
+#undef RP_ADC_USE_ADC1
+#define RP_ADC_USE_ADC1 TRUE
+
+#undef RP_IRQ_ADC1_PRIORITY
+#define RP_IRQ_ADC1_PRIORITY 2
```
Comment (tzarc) [inline keyboards/handwired/onekey/rp2040/mcuconf.h:18]: Shall we move this into the default?

Comment (sigprof) [reply keyboards/handwired/onekey/rp2040/mcuconf.h:18]: Probably yes (we have 3 such files though).

**Comment by sigprof** [comment]
I changed the priority to 3 to match the value used by ChibiOS tests:
```
lib/chibios-contrib/testhal/RP/RP2040/RT-RP2040-PICO-ADC/cfg/mcuconf.h:#define RP_IRQ_ADC1_PRIORITY                3
lib/chibios-contrib/testhal/RP/RP2040/RT-RP2040-PICO-I2C-24AA01/cfg/mcuconf.h:#define RP_IRQ_ADC1_PRIORITY                3
platforms/chibios/boards/GENERIC_PROMICRO_RP2040/configs/mcuconf.h:#define RP_IRQ_ADC1_PRIORITY                3
platforms/chibios/boards/GENERIC_RP_RP2040/configs/mcuconf.h:#define RP_IRQ_ADC1_PRIORITY                3
platforms/chibios/boards/QMK_PM2040/configs/mcuconf.h:#define RP_IRQ_ADC1_PRIORITY                3
```
This should still work, but I can't actually test this at the moment.

File: docs/adc_driver.md [hunk 2]
```diff
@@ -121,6 +123,21 @@ Also note that the F0 and F3 use different numbering schemes. The F0 has a singl
...
+
+RP2040 has only a single ADC (`ADCD1` in ChibiOS); in the QMK API the index for that ADC is 0.
+
+|Channel|Pin                |
+|-------|-------------------|
+|0      |`GP26`             |
+|1      |`GP27`             |
+|2      |`GP28`             |
+|3      |`GP29`             |
+|4      |Temperature sensor¹|
```
Comment (fauxpark) [inline docs/adc_driver.md:136]:
```suggestion
|4      |Temperature sensor*|
```

File: docs/adc_driver.md [hunk 3]
```diff
@@ -121,6 +123,21 @@ Also note that the F0 and F3 use different numbering schemes. The F0 has a singl
...
+
+
+<sup>¹ The temperature sensor is disabled by default and needs to be enabled by the RP2040-specific function: `adcRPEnableTS(&ADCD1)`.  The ADC must be initialized before calling that function; an easy way to ensure that is to perform a dummy conversion.</sup>
```
Comment (fauxpark) [inline docs/adc_driver.md:139]:
```suggestion
<sup>* The temperature sensor is disabled by default and needs to be enabled by the RP2040-specific function: `adcRPEnableTS(&ADCD1)`.  The ADC must be initialized before calling that function; an easy way to ensure that is to perform a dummy conversion.</sup>
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by tzarc** [comment]
> This should still work, but I can't actually test this at the moment.

Works fine.

**Review by KarlK90 — APPROVED** [review]
Could you update the rp2040 platform dev line about the ADC driver? https://github.com/qmk/qmk_firmware/blob/master/docs/platformdev_rp2040.md#L7 The other changes LGTM.

**Comment by sigprof** [comment]
Also I suppose that the `QMK_PM2040` board should now have `RP_ADC_USE_ADC1` and `HAL_USE_ADC` enabled for the benefit of converters?

What about the `GENERIC_PROMICRO_RP2040` board (which is apparently the default one selected for `"processor": "RP2040"`?  Apparently that board now has `RP_SIO_USE_UART0`, `RP_SPI_USE_SPI0`, `RP_I2C_USE_I2C1` enabled, but the corresponding HAL options are not enabled anywhere; it probably should have `RP_ADC_USE_ADC1` enabled too?

**Comment by sigprof** [comment]
Looks like it is possible to get something like `qmk compile -kb handwired/onekey/promicro -km adc -e CONVERT_TO=elite_pi` to build, but it requires an additional change to `analog.c`, because the `GPxx` defines are not available when a converter is used:
```diff
diff --git a/platforms/chibios/drivers/analog.c b/platforms/chibios/drivers/analog.c
index 84ca99be3c..a8ce21bb6d 100644
--- a/platforms/chibios/drivers/analog.c
+++ b/platforms/chibios/drivers/analog.c
@@ -241,10 +241,10 @@ __attribute__((weak)) adc_mux pinToMux(pin_t pin) {
         // STM32F103x[C-G] in 144-pin packages also have analog inputs on F6...F10, but they are on ADC3, and the
         // ChibiOS ADC driver for STM32F1xx currently supports only ADC1, therefore these pins are not usable.
 #elif defined(RP2040)
-        case GP26: return TO_MUX(0, 0);
-        case GP27: return TO_MUX(1, 0);
-        case GP28: return TO_MUX(2, 0);
-        case GP29: return TO_MUX(3, 0);
+        case 26U: return TO_MUX(0, 0);
+        case 27U: return TO_MUX(1, 0);
+        case 28U: return TO_MUX(2, 0);
+        case 29U: return TO_MUX(3, 0);
 #endif
     }


```
Would this be acceptable?  (If it would, then the STM32 cases above should probably be changed to use something like `PAL_LINE(GPIOA, 0)` too.)

**Comment by drashna** [comment]
would adding `pin_defs.h` work here, as well?

**Comment by sigprof** [comment]
The converter support overrides `_pin_defs.h` with the converter version that defines only the pin names for the source MCU, so the pin names for the conversion target MCU are not available (and in general they may even conflict with the pin names for the source MCU, so there is no way to get them both). So apparently if the driver code needs to refer to some specific target MCU pins (as is the case for `pinToMux()` in the ADC driver), the only option to make that kind of code work with converters is to use something else instead of the QMK pin names.

Most other drivers don't have this problem, because the pin names are passed externally in defines (which, BTW, would contain the source pin names in case a converter is used), and the only places where explicit pin names are used in the driver code are some defaults which the converters override anyway.

**Comment by KarlK90** [comment]
> Also I suppose that the `QMK_PM2040` board should now have `RP_ADC_USE_ADC1` and `HAL_USE_ADC` enabled for the benefit of converters?

To be inline with the expectation that a converter "just works" this should be done.

> What about the `GENERIC_PROMICRO_RP2040` board (which is apparently the default one selected for `"processor": "RP2040"`)? Apparently that board now has `RP_SIO_USE_UART0`, `RP_SPI_USE_SPI0`, `RP_I2C_USE_I2C1` enabled, but the corresponding HAL options are not enabled anywhere; it probably should have `RP_ADC_USE_ADC1` enabled too?

Yes, I think the ADC peripheral should be enabled as well without activating the HAL driver.

**Comment by KarlK90** [comment]
The whole `analog.c/adc` handling could probably see a overhaul, but I think the way it is handled in the scope of RP2040 support is fine for now.

**Comment by tzarc** [comment]
> The whole `analog.c/adc` handling could probably see a overhaul, but I think the way it is handled in the scope of RP2040 support is fine for now.

It's been ripe for replacing `SRC += analog.c` with `ANALOG_ENABLE = yes` for quite some time. Adding the defines for enabling board-specific ADCs at that point would be relatively trivial.

**Comment by zvecr** [comment]
From my perspective, I dont care too much if converters fail. Long term it would be expected to work, but it shouldnt be a blocker for getting ADC support in.

**Comment by sigprof** [comment]
Actually the converters should work with the latest code here.  The only issue is that some keyboards may use analog inputs on Pro Micro/Elite-C pins outside of the `F4`…`F7` range, and compiling that code with `CONVERT_TO=some_rp2040_controller` would silently give a broken firmware, because the unsupported pin usage could be detected only at runtime.  But I don't see any good way around that (even if we add some `analogReadPin()` wrapper using `__builtin_constant_p()` to have a compile time check if the argument is constant, that won't help in cases when the pin is passed indirectly, such as when using the joystick feature).  Maybe that's a mostly theoretical issue though (apparently all existing boards that use the joystick feature use pins from the `F4`…`F7` range, and boards that use other pins might actually be using a custom PCB, not a replaceable controller).

**Comment by KarlK90** [comment]
So this sounds like the PR is ready to go in?! The refactoring should be done in a separate PR imho.

**Comment by tzarc** [comment]
Can iterate on later PRs if required.

**Pull request merged**